### ISSUE TITLE
Feature request: Support for Deno's lsp

### ISSUE THREAD

**Issue by oxinosg** [emacs-lsp/lsp-mode#2235]
[Deno: A secure runtime for JavaScript and TypeScript.](https://deno.land/)

Because of Deno's module system the current JS/TS language servers cannot be used, but the team developed a VS Code plugin that uses lsp [vscode_deno](https://github.com/denoland/vscode_deno)

**Comment by evanlouie** [comment]
Deno 1.6 just released a built in LSP https://deno.land/posts/v1.6#built-in-deno-language-server

**Comment by XO39** [comment]
@yyoncho Sorry for the duplication.

Any plan for this?

I tried to mess around and did this (basically copied from `flow-ls` code in the `lsp-javascript.el` file):
```elisp
(require 'lsp-mode)

(define-derived-mode ts-mode typescript-mode "TS"
  "Test typescript-mode for javascript."
  )
(add-to-list 'auto-mode-alist '("\\.ts?\\'" . ts-mode))

(add-to-list 'lsp-language-id-configuration '(ts-mode . "typescript"))

(defgroup lsp-deno nil
  "LSP support for using Deno builtin Language Server."
  :group 'lsp-mode
  :link '(url-link "https://deno.land"))

(defcustom lsp-clients-deno-server "deno"
  "The deno executable to use.
Leave as just the executable name to use the default behavior of
finding the executable with variable `exec-path'."
  :group 'lsp-deno
  :risky t
  :type 'file)

(defcustom lsp-clients-deno-server-args '("lsp")
  "Extra arguments for the deno-language-server language server."
  :group 'lsp-deno
  :risky t
  :type '(repeat string))

(defcustom lsp-clients-deno-log-verbosity "info"
  "The server log verbosity."
  :group 'lsp-deno
  :type 'string)

(defun lsp-clients-deno-activate-p (file-name _mode)
  "Check if the Deno language server should be enabled for a particular FILE-NAME and MODE."
  (and (derived-mode-p 'js-mode 'web-mode 'js2-mode 'rjsx-mode 'ts-mode)
       (not (derived-mode-p 'json-mode))))

(lsp-register-client
 (make-lsp-client :new-connection
                  (lsp-stdio-connection (lambda ()
                                          (cons lsp-clients-deno-server
                                                lsp-clients-deno-server-args)))
                  :priority 0
                  :activation-fn 'lsp-clients-deno-activate-p
                  :server-id 'deno-ls))

```

 But, although the new client `deno-ls` fires whenever I open a TypeScript file, all functions and features of lsp-mode are not there, no popups, no ui nothing!

This is from `*lsp-log*`
```
Found the following clients for /Users/xo/sandbox/deno/deno.ts: (server-id deno-ls, priority 0), (server-id ts-ls, priority -2)
The following clients were selected based on priority: (server-id deno-ls, priority 0)
This server does not support foldingRangeProvider
```

And this is from thei `*deno-ls::stderr*`:
```
Starting Deno language server...
  version: 1.7.0 (release, x86_64-apple-darwin)
Connected to "emacs" GNU Emacs 27.1 (build 1, x86_64-apple-darwin18.7.0, NS appkit-1671.60 Version 10.14.6 (Build 18G95))
 of 2020-08-12
language server initialized
Server ready.
```

What am I missing?

**Comment by mkhattab** [comment]
@XO39 

I'm having trouble as well. I'm not familiar with the LSP specs, but I wrote a small script in typescript to determine if there's an issue with the `deno lsp` command or `lsp-mode`. 

**TL;DR** it looks like an issue with `deno lsp` and it seems broken and unfinished, unless I'm missing something, which I probably am.

```typescript
import * as cp from 'child_process';
import * as rpc from 'vscode-languageserver-protocol/node';
import * as fs from 'fs';

const ROOT_PATH_URI = 'file:///path/to/some/dir';
const FILE_PATH = '/path/to/some/dir/example.ts';

async function main() {
  console.log('starting lsp client...');

  const childProcess = cp.spawn('deno', ['lsp']);

  // Uncomment to verify script works with the standard typescript lsp server
  // const childProcess = cp.spawn('typescript-language-server', ['--stdio']);
  
  childProcess.stdout.on('data', (data) => console.log(`debug stdout: ${data}\n`));
  childProcess.stderr.on('data', (data) => console.log(`debug stderr: ${data}\n`));

  const connection = rpc.createProtocolConnection(
    new rpc.StreamMessageReader(childProcess.stdout),
    new rpc.StreamMessageWriter(childProcess.stdin)
  )
  connection.listen();

  const init: rpc.InitializeParams = {
    rootUri: ROOT_PATH_URI,
    processId: 1,
    capabilities: {},
    workspaceFolders: null,
  };

  const initResult = await connection.sendRequest(rpc.InitializeRequest.type, init);
  console.log('debug response: ', initResult);
  console.log('debug response: ', initResult.capabilities.completionProvider?.triggerCharacters);

  const fileUri = `file://${FILE_PATH}`;
  const fileContents = fs.readFileSync(FILE_PATH).toString();

  const didOpenParams: rpc.DidOpenTextDocumentParams = {
    textDocument: {
      uri: fileUri,
      languageId: 'typescript',
      version: 1,
      text: fileContents,
    }
  };
  connection.sendNotification(rpc.DidOpenTextDocumentNotification.type, didOpenParams);

  const completionParams: rpc.CompletionParams = {
    textDocument: {
      uri: fileUri,
    },
    position: {
      line: 1,
      character: 11,
    },
    context: {
      triggerKind: rpc.CompletionTriggerKind.Invoked,
    },
  };
  const completionResult = await connection.sendRequest(rpc.CompletionRequest.type, completionParams);
  console.log('debug response: ', completionResult);
}

main();
```

Output
```
starting lsp client...
debug stderr: Starting Deno language server...
  version: 1.7.1 (release, x86_64-apple-darwin)


debug stderr: language server initialized


debug stdout: Content-Length: 475

{"jsonrpc":"2.0","result":{"capabilities":{"textDocumentSync":{"openClose":true,"change":2,"save":{}},"hoverProvider":true,"completionProvider":{"triggerCharacters":[".","\"","'","`","/","@","<","#"]},"definitionProvider":true,"implementationProvider":true,"referencesProvider":true,"documentHighlightProvider":true,"documentFormattingProvider":true,"renameProvider":true},"serverInfo":{"name":"deno-language-server","version":"1.7.1 (release, x86_64-apple-darwin)"}},"id":0}

initializeResult: {
  capabilities: {
    textDocumentSync: { openClose: true, change: 2, save: {} },
    hoverProvider: true,
    completionProvider: { triggerCharacters: [Array] },
    definitionProvider: true,
    implementationProvider: true,
    referencesProvider: true,
    documentHighlightProvider: true,
    documentFormattingProvider: true,
    renameProvider: true
  },
  serverInfo: {
    name: 'deno-language-server',
    version: '1.7.1 (release, x86_64-apple-darwin)'
  }
}
debug stdout: Content-Length: 38



debug stdout: {"jsonrpc":"2.0","result":null,"id":1}

completionResult: null
```

**Comment by mkhattab** [comment]
If anyone is stuck on this and wants to use deno + emacs, using the `typescript-deno-plugin` is a decent work around for now. Just install (globally) and add the plugin to your `tsconfig.json` as instructed [here](https://github.com/justjavac/typescript-deno-plugin#configuration).

**Comment by mkhattab** [comment]
Managed to get `deno lsp` to work. You need to set `"enable": true` in `initializationOptions`, which is `false` by default.

E.g.

```emacs
(lsp-register-client
 (make-lsp-client :new-connection
                  (lsp-stdio-connection (lambda ()
                                          (cons lsp-clients-deno-server
                                                lsp-clients-deno-server-args)))
                  :priority 0
                  :activation-fn 'lsp-clients-deno-activate-p
                  :initialization-options '(:enable t)
                  :server-id 'deno-ls))
```

See [this](https://github.com/denoland/vscode_deno#configuration) for more configuration options you can set in via `:initialization-options`

**Comment by yyoncho** [comment]
> ```emacs-lisp
>   :initialization-options '(:enable t)
> ```

The explanation:

> deno.enable: Controls if the Deno Language Server is enabled. When enabled, the extension will disable the built-in VSCode 
> JavaScript and TypeScript language services, and will use the Deno Language Server (deno lsp) instead. boolean, default 

VScode extension mechanisms in action! (3 lines of rofl emojis).

@mkhattab willing to make a PR?

**Comment by mkhattab** [comment]
@yyoncho Yes, I think I can take a shot at this.

**Comment by mkhattab** [comment]
@yyoncho 

Let's say we have a variable `lsp-deno-enabled`, if it's `non-nil` do you think the Deno LSP client should take priority over the Typescript LSP client? 

This behavior is similar to how the Flow LSP client is [configured](https://github.com/emacs-lsp/lsp-mode/blob/master/clients/lsp-javascript.el#L192). And it's also similar to how the [vscode_deno plugin](https://github.com/denoland/vscode_deno#configuration) behaves as well.

Or...should we perhaps let users explicitly specify if they want to use the Deno LSP client via `lsp-enabled-clients` as stated in the [docs](https://emacs-lsp.github.io/lsp-mode/page/faq/)?

Wanted to get your guidance on this before I start on the PR.

**Comment by yyoncho** [comment]
@mkhattab what we do is set a default priority (in this case I would say the priority of deno should be the lowest) and then user will use lsp-enabled-clients if they want to use it in a particular project or they will disable the other servers via lsp-disabled-clients globally. FWIW the vscode solution to disable deno lsp is bizarre because it starts the server although it is not needed at all and the server reports no capabilities...

**Comment by mkhattab** [comment]
@yyoncho Got it, thanks!

**Comment by yyoncho** [comment]
Fixed by #2607

**Comment by HarunAt1Password** [comment]
As a new emacs user, how do I even get this up and running ? The documentation just says "install lsp" so I installed deno but how do I get Emacs to use Deno over TS/JS ? There should be more detail in the actual documentation .. I couldn't find any resources online either

**Comment by yyoncho** [comment]
> how do I get Emacs to use Deno over TS/JS ? 

https://emacs-lsp.github.io/lsp-mode/page/faq/#i-have-multiple-language-servers-registered-for-language-foo-which-one-will-be-used-when-opening-a-project

**Comment by HarunAt1Password** [comment]
> https://emacs-lsp.github.io/lsp-mode/page/faq/#i-have-multiple-language-servers-registered-for-language-foo-which-one-will-be-used-when-opening-a-project

So all I gotta do is ?


```
(use-package lsp-mode
	:commands (lsp lsp-deferred)
        :add-on t)
```

**Comment by yyoncho** [comment]
@mkhattab you should use `lsp-disabled-clients` to disable `ts-ls`

**Comment by yyoncho** [comment]
Something like: `(setq lsp-disabled-clients '(ccls rls mspyls eslint angular-ls nimls))`

**Comment by HarunAt1Password** [comment]
oof, I don't want to disable it tho! lol, I want to keep it available- i'm just really just want deno to run when I am in a project that is a "deno workspace".

**Comment by yyoncho** [comment]
ATM the option to do that is to use dir local with lsp-disabled-clients in the projects you want to use. Or write custom elisp function to do the checks for you.

**Comment by HarunAt1Password** [comment]
@yyoncho thanks for explaining it to me in more detail -- really appreciate it!

**Comment by CadeMichael** [comment]
would there be an easy way to get the deno lsp to only activate for typescript files and use the standard ts-ls for js?

**Comment by yyoncho** [comment]
@CadeMichael dir locals can be per major mode, so you can achieve that using lsp-disabled-clients. You may utilize `M-x add-dir-local-variable`

**Comment by CadeMichael** [comment]
@yyoncho Thank you!

**Comment by CadeMichael** [comment]
I was able to get the desired behavior with this, albeit clunky solution, 

```
(use-package typescript-mode
  :ensure t
  :hook (typescript-mode . (lambda ()
			     (setq lsp-disabled-clients '(ts-ls))
			     (lsp-deferred))))
(add-hook 'js-mode-hook (lambda ()
			  (setq lsp-disabled-clients '())
			  (lsp-deferred)))

```

**Comment by yyoncho** [comment]
@CadeMichael this is also an acceptable solution. (Typically the variable is let bound to achieve the same)

**Comment by Adek06** [comment]
Hi everyone, I'm new to deno and I'm trying to install the deno server, but I can't find it in the list. What should I do?

![image](https://user-images.githubusercontent.com/25602645/224502386-fc1ad21e-55c0-48f2-8b6e-7d118f76da11.png)
![image](https://user-images.githubusercontent.com/25602645/224502397-fcad23b2-8191-48b5-8493-393070da9ca5.png)
![image](https://user-images.githubusercontent.com/25602645/224502533-a93405db-9cca-4b95-83ab-eaf7b73a59fa.png)

**Comment by numkem** [comment]
@Adek06 just install deno itself. It’s included.

**Comment by Adek06** [comment]
@numkem Thanks，it's work!

**Comment by ranjanified** [comment]
> > how do I get Emacs to use Deno over TS/JS ?
> 
> https://emacs-lsp.github.io/lsp-mode/page/faq/#i-have-multiple-language-servers-registered-for-language-foo-which-one-will-be-used-when-opening-a-project

Thank You 🙏. Was running into this due to `lsp-svelte` and `lsp-javascript` (`'deno-ls`). Pursued `.dir-locals.el` way to work around it for now. 

Will it be possible for `lsp` to accept this `lsp-enabled-clients` as arguments ? Just a thought.

**Comment by yyoncho** [comment]
> Will it be possible for `lsp` to accept this `lsp-enabled-clients` as arguments ?

@ranjanified lsp is for interactive use but you can still pass it using a dynamic binding. Something like:

```
(let ((lsp-enabled-clients '(foo-bar)))
   (lsp))
```

**Comment by ranjanified** [comment]
> > Will it be possible for `lsp` to accept this `lsp-enabled-clients` as arguments ?
> 
> @ranjanified lsp is for interactive use but you can still pass it using a dynamic binding. Something like:
> 
> ```
> (let ((lsp-enabled-clients '(foo-bar)))
>    (lsp))
> ```

Thank You yet again 🙏. This one is sensible and I'm gonna stick to this way for now, as this achieves what I need, and a fit in the way I am calling `lsp-mode`.

**Comment by zhengpd** [comment]
This is my setup to use deno lsp only in deno projects:

```elisp
  ;; Use deno-ls in deno projects and ts-ls for general typescript files
  (add-hook 'typescript-ts-mode-hook
	    #'(lambda nil
		(if (or (locate-dominating-file (buffer-file-name) "deno.json")
			(locate-dominating-file (buffer-file-name) "deno.jsonc"))
		    (setq-local lsp-enabled-clients '(deno-ls))
		  (setq-local lsp-enabled-clients '(ts-ls)))
		(call-interactively 'lsp)))
```
